pkg/types/validators.go

```diff
@@ -1,28 +1,52 @@
 package types
 
 import (
+	"time"
+
 	"github.com/mt-sre/addon-metadata-operator/api/v1alpha1"
 	"github.com/operator-framework/operator-registry/pkg/registry"
 )
 
-type ValidateFunc func(mb MetaBundle) ValidatorResult
-
+// Validator encapsulates display fields and implementation for
+// tasks performed against MetaBundles to validate their data.
 type Validator struct {
 	Name        string
 	Code        string
 	Description string
-	Runner      ValidateFunc
+	Runner      ValidatorRunner
 	// Optional: override defaults from RetryMiddleware
 	RetryCount int
 	// Optional: override defaults from RetryMiddleware
-	RetryDelaySeconds int
+	RetryDelayer Delayer
 }
 
-// WithRunner - Needs to create a new validator otherwise the memory address of
-// the original Runner is lost, and it causes infinite loop scenarios (e.g.: RetryMiddleware)
-func (v Validator) WithRunner(fn ValidateFunc) Validator {
-	v.Runner = fn
-	return v
+// Validate runs a validator's Runner against a MetaBundle.
+func (v *Validator) Validate(mb MetaBundle) ValidatorResult {
+	return v.Runner.Run(mb)
+}
+
+// ValidatorRunner describes objects which run validations against
+// a MetaBundle and produce a ValidatorResult. This is analogous to
+// the http.Handler interface.
+type ValidatorRunner interface {
+	// Run validates a MetaBundle and produces a ValidatorResult
+	// indicating whether the validations succedded, failed, or errored out.
+	// This is analogous to Handler.ServeHTTP
+	Run(MetaBundle) ValidatorResult
+}
+
+// ValidateFunc wraps functions which validate MetaBundles.
+type ValidateFunc func(mb MetaBundle) ValidatorResult
+
+func (f ValidateFunc) Run(mb MetaBundle) ValidatorResult {
+	return f(mb)
+}
```

Comment:
Do we still need this?

Reply:
Yes, so all the questions related to typecasting `ValidateFunc` can be answered here:

Basically a `func(mb MetaBundle) ValidatorResult` cannot be given a `Run` method and will not satisfy the `ValidatorRunner` interface. So we need this additional abstraction to cast our validation functions as instances of `ValidateFunc` to satisfy the interface. It's definitely a funky Go thing.